### PR TITLE
Using Optimization.jl

### PR DESCRIPTION
**Pull request by ba2tripleO**

### CODE REVIEW

File: docs/src/neural_ode/neural_gde.md [hunk 1]
```diff
@@ -293,20 +294,20 @@ st = st |> device
...
+Finally, we use the package `Optimization.solve` to learn the parameters `ps`. We run the training loop for `epochs` number of iterations.
 
 ```@example graphneuralode
 for _ in 1:epochs
-    (l,st), back = pullback(p->loss(X, ytrain, train_mask, model, p, st), ps)
-    gs = back((one(l), nothing))[1]
-    st_opt, ps = Optimisers.update(st_opt, ps, gs)
-    @show eval_loss_accuracy(X, y, val_mask, model, ps, st)
+	optprob = Optimization.OptimizationProblem(optf, ps)
+	ps = Optimization.solve(optprob, opt).minimizer
```
Comment (YichengDWu) [inline docs/src/neural_ode/neural_gde.md]:
```julia
ERROR: The number of maxiters has to be a non-negative and non-zero number.
Stacktrace:
```

File: docs/src/neural_ode/neural_gde.md [hunk 2]
```diff
@@ -9,9 +9,10 @@ In this tutorial we will use Graph Differential Equations (GDEs) to perform clas
 using GraphNeuralNetworks, DifferentialEquations
 using DiffEqFlux: NeuralODE
 using GraphNeuralNetworks.GNNGraphs: normalized_adjacency
-using Lux, NNlib, Optimisers, Zygote, Random, ComponentArrays
+using Lux, NNlib, Zygote, Random, ComponentArrays
 using Lux: AbstractExplicitLayer, glorot_normal, zeros32
 import Lux: initialparameters, initialstates
+import Optimization, OptimizationFlux
```
Comment (YichengDWu) [inline docs/src/neural_ode/neural_gde.md]: using Optimization
using OptimizationFlux: ADAM

File: docs/src/neural_ode/neural_gde.md [hunk 3]
```diff
@@ -293,20 +294,20 @@ st = st |> device
...
 for _ in 1:epochs
-    (l,st), back = pullback(p->loss(X, ytrain, train_mask, model, p, st), ps)
-    gs = back((one(l), nothing))[1]
-    st_opt, ps = Optimisers.update(st_opt, ps, gs)
-    @show eval_loss_accuracy(X, y, val_mask, model, ps, st)
+	optprob = Optimization.OptimizationProblem(optf, ps)
+	ps = Optimization.solve(optprob, opt).minimizer
+	@show eval_loss_accuracy(X, y, val_mask, model, ps, st)
 end
 ```
```
Comment (YichengDWu) [inline docs/src/neural_ode/neural_gde.md:320]: Change the output here

**Review by YichengDWu — CHANGES_REQUESTED** [review]

File: docs/src/neural_ode/neural_gde.md (hunk 1 above)
Comment (YichengDWu) [reply docs/src/neural_ode/neural_gde.md]: Still not working
```julia
ERROR: This object is not a GPU array
Stacktrace:
  [1] error(s::String)
    @ Base .\error.jl:33
  [2] backend(#unused#::Type)
    @ GPUArrays C:\Users\Luffy\.julia\packages\GPUArrays\Zecv7\src\device\execution.jl:15
  [3] backend(x::ComponentVector{Float32, CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}, Tuple{Axis{(layer_1 = ViewAxis(1:22944, Axis(weight = ViewAxis(1:22928, ShapedAxis((16, 1433), NamedTuple())), bias = ViewAxis(22929:22944, ShapedAxis((16, 1), NamedTuple())))), layer_2 = ViewAxis(22945:23488, Axis(layer_1 = ViewAxis(1:272, Axis(weight = ViewAxis(1:256, ShapedAxis((16, 16), NamedTuple())), bias = ViewAxis(257:272, ShapedAxis((16, 1), NamedTuple())))), layer_2 = ViewAxis(273:544, Axis(weight = ViewAxis(1:256, ShapedAxis((16, 16), NamedTuple())), bias = ViewAxis(257:272, ShapedAxis((16, 1), NamedTuple())))))), layer_3 = 23489:23488, layer_4 = ViewAxis(23489:23607, Axis(weight = ViewAxis(1:112, ShapedAxis((7, 16), NamedTuple())), bias = ViewAxis(113:119, ShapedAxis((7, 1), NamedTuple())))))}}})
    @ GPUArrays C:\Users\Luffy\.julia\packages\GPUArrays\Zecv7\src\device\execution.jl:16
  [4] _copyto!
    @ C:\Users\Luffy\.julia\packages\GPUArrays\Zecv7\src\host\broadcast.jl:73 [inlined]
  [5] materialize!
    @ C:\Users\Luffy\.julia\packages\GPUArrays\Zecv7\src\host\broadcast.jl:51 [inlined]
  [6] materialize!(dest::ComponentVector{Float32, CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}, Tuple{Axis{(layer_1 = ViewAxis(1:22944, Axis(weight = ViewAxis(1:22928, ShapedAxis((16, 1433), NamedTuple())), bias = ViewAxis(22929:22944, ShapedAxis((16, 1), NamedTuple())))), layer_2 = ViewAxis(22945:23488, Axis(layer_1 = ViewAxis(1:272, Axis(weight = ViewAxis(1:256, ShapedAxis((16, 16), NamedTuple())), bias = ViewAxis(257:272, ShapedAxis((16, 1), NamedTuple())))), layer_2 = ViewAxis(273:544, Axis(weight = ViewAxis(1:256, ShapedAxis((16, 16), NamedTuple())), bias = ViewAxis(257:272, ShapedAxis((16, 1), NamedTuple())))))), layer_3 = 23489:23488, layer_4 = ViewAxis(23489:23607, Axis(weight = ViewAxis(1:112, ShapedAxis((7, 16), NamedTuple())), bias = ViewAxis(113:119, ShapedAxis((7, 1), NamedTuple())))))}}}, bc::Base.Broadcast.Broadcasted{CUDA.CuArrayStyle{1}, Nothing, typeof(identity), Tuple{ComponentVector{Float32, CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}, Tuple{Axis{(layer_1 = ViewAxis(1:22944, Axis(weight = ViewAxis(1:22928, ShapedAxis((16, 1433), NamedTuple())), bias = ViewAxis(22929:22944, ShapedAxis((16, 1), NamedTuple())))), layer_2 = ViewAxis(22945:23488, Axis(layer_1 = ViewAxis(1:272, Axis(weight = ViewAxis(1:256, ShapedAxis((16, 16), NamedTuple())), bias = ViewAxis(257:272, ShapedAxis((16, 1), NamedTuple())))), layer_2 = ViewAxis(273:544, Axis(weight = ViewAxis(1:256, ShapedAxis((16, 16), NamedTuple())), bias = ViewAxis(257:272, ShapedAxis((16, 1), NamedTuple())))))), layer_3 = 23489:23488, layer_4 = ViewAxis(23489:23607, Axis(weight = ViewAxis(1:112, ShapedAxis((7, 16), NamedTuple())), bias = ViewAxis(113:119, ShapedAxis((7, 1), NamedTuple())))))}}}}})
    @ Base.Broadcast .\broadcast.jl:868
  [7] (::Optimization.var"#117#127"{Optimization.var"#116#126"{OptimizationFunction{true, Optimization.AutoZygote, var"#19#20", Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing}, SciMLBase.NullParameters}})(::ComponentVector{Float32, CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}, Tuple{Axis{(layer_1 = ViewAxis(1:22944, Axis(weight = ViewAxis(1:22928, ShapedAxis((16, 1433), NamedTuple())), bias = ViewAxis(22929:22944, ShapedAxis((16, 1), NamedTuple())))), layer_2 = ViewAxis(22945:23488, Axis(layer_1 = ViewAxis(1:272, Axis(weight = ViewAxis(1:256, ShapedAxis((16, 16), NamedTuple())), bias = ViewAxis(257:272, ShapedAxis((16, 1), NamedTuple())))), layer_2 = ViewAxis(273:544, Axis(weight = ViewAxis(1:256, ShapedAxis((16, 16), NamedTuple())), bias = ViewAxis(257:272, ShapedAxis((16, 1), NamedTuple())))))), layer_3 = 23489:23488, layer_4 = ViewAxis(23489:23607, Axis(weight = ViewAxis(1:112, ShapedAxis((7, 16), NamedTuple())), bias = ViewAxis(113:119, ShapedAxis((7, 1), NamedTuple())))))}}}, ::ComponentVector{Float32, CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}, Tuple{Axis{(layer_1 = ViewAxis(1:22944, Axis(weight = ViewAxis(1:22928, ShapedAxis((16, 1433), NamedTuple())), bias = ViewAxis(22929:22944, ShapedAxis((16, 1), NamedTuple())))), layer_2 = ViewAxis(22945:23488, Axis(layer_1 = ViewAxis(1:272, Axis(weight = ViewAxis(1:256, ShapedAxis((16, 16), NamedTuple())), bias = ViewAxis(257:272, ShapedAxis((16, 1), NamedTuple())))), layer_2 = ViewAxis(273:544, Axis(weight = ViewAxis(1:256, ShapedAxis((16, 16), NamedTuple())), bias = ViewAxis(257:272, ShapedAxis((16, 1), NamedTuple())))))), layer_3 = 23489:23488, layer_4 = ViewAxis(23489:23607, Axis(weight = ViewAxis(1:112, ShapedAxis((7, 16), NamedTuple())), bias = ViewAxis(113:119, ShapedAxis((7, 1), NamedTuple())))))}}})
    @ Optimization C:\Users\Luffy\.julia\packages\Optimization\C5BBJ\src\function\zygote.jl:32
  [8] macro expansion
    @ C:\Users\Luffy\.julia\packages\OptimizationFlux\cpWyO\src\OptimizationFlux.jl:32 [inlined]
  [9] macro expansion
    @ C:\Users\Luffy\.julia\packages\Optimization\C5BBJ\src\utils.jl:35 [inlined]
 [10] __solve(prob::OptimizationProblem{true, OptimizationFunction{true, Optimization.AutoZygote, var"#19#20", Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing}, ComponentVector{Float32, CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}, Tuple{Axis{(layer_1 = ViewAxis(1:22944, Axis(weight = ViewAxis(1:22928, ShapedAxis((16, 1433), NamedTuple())), bias = ViewAxis(22929:22944, ShapedAxis((16, 1), NamedTuple())))), layer_2 = ViewAxis(22945:23488, Axis(layer_1 = ViewAxis(1:272, Axis(weight = ViewAxis(1:256, ShapedAxis((16, 16), NamedTuple())), bias = ViewAxis(257:272, ShapedAxis((16, 1), NamedTuple())))), layer_2 = ViewAxis(273:544, Axis(weight = ViewAxis(1:256, ShapedAxis((16, 16), NamedTuple())), bias = ViewAxis(257:272, ShapedAxis((16, 1), NamedTuple())))))), layer_3 = 23489:23488, layer_4 = ViewAxis(23489:23607, Axis(weight = ViewAxis(1:112, ShapedAxis((7, 16), NamedTuple())), bias = ViewAxis(113:119, ShapedAxis((7, 1), NamedTuple())))))}}}, SciMLBase.NullParameters, Nothing, Nothing, Nothing, Nothing, Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}}}, opt::ADAM, data::Base.Iterators.Cycle{Tuple{Optimization.NullData}}; maxiters::Int64, callback::Function, progress::Bool, save_best::Bool, kwargs::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
    @ OptimizationFlux C:\Users\Luffy\.julia\packages\OptimizationFlux\cpWyO\src\OptimizationFlux.jl:30
 [11] #solve#492
    @ C:\Users\Luffy\.julia\packages\SciMLBase\koNdH\src\solve.jl:71 [inlined]
 [12] top-level scope
    @ REPL[67]:1
```
Also I'm sure if we need a for loop anymore. It seems the iteration is defined inside `solve`. In this case you need to put  `eval_loss_accuracy` inside a callback function

File: docs/src/neural_ode/neural_gde.md (hunk 3 above)
Comment (YichengDWu) [reply docs/src/neural_ode/neural_gde.md:320]: Sorry I meant add the printed output

File: docs/src/neural_ode/neural_gde.md [hunk 4]
```diff
@@ -293,21 +304,18 @@ st = st |> device
...
-for _ in 1:epochs
-    (l,st), back = pullback(p->loss(X, ytrain, train_mask, model, p, st), ps)
-    gs = back((one(l), nothing))[1]
-    st_opt, ps = Optimisers.update(st_opt, ps, gs)
-    @show eval_loss_accuracy(X, y, val_mask, model, ps, st)
-end
+ps = Optimization.solve(optprob, opt, maxiters=1000, callback=callback).minimizer
 ```
 
 ## Expected Output
```
Comment (YichengDWu) [inline docs/src/neural_ode/neural_gde.md:321]: add the printed output here

File: docs/src/neural_ode/neural_gde.md [hunk 5]
```diff
@@ -293,21 +304,18 @@ st = st |> device
...
+Finally, we use the package `Optimization.solve` to learn the parameters `ps`. We run the training loop for `epochs` number of iterations.
 
 ```@example graphneuralode
-for _ in 1:epochs
-    (l,st), back = pullback(p->loss(X, ytrain, train_mask, model, p, st), ps)
-    gs = back((one(l), nothing))[1]
-    st_opt, ps = Optimisers.update(st_opt, ps, gs)
-    @show eval_loss_accuracy(X, y, val_mask, model, ps, st)
-end
+ps = Optimization.solve(optprob, opt, maxiters=1000, callback=callback).minimizer
```
Comment (YichengDWu) [inline docs/src/neural_ode/neural_gde.md:318]: We don't need to return `ps` anymore

`Optimization.solve(optprob, opt, maxiters=1000, callback=callback)`

File: docs/src/neural_ode/neural_gde.md [hunk 6]
```diff
@@ -293,21 +304,18 @@ st = st |> device
...
+opt = ADAM(0.01f0)
+loss_clr(p) = loss(X, ytrain, train_mask, model, p, st)[1]
+optf = Optimization.OptimizationFunction((x,p)->loss_clr(x), Optimization.AutoZygote())
+optprob = Optimization.OptimizationProblem(optf, ps)
 ```
 
 ## Training Loop
 
-Finally, we use the package `Optimisers` to learn the parameters `ps`. We run the training loop for `epochs` number of iterations.
+Finally, we use the package `Optimization.solve` to learn the parameters `ps`. We run the training loop for `epochs` number of iterations.
```
Comment (YichengDWu) [inline docs/src/neural_ode/neural_gde.md:315]: Should `maxiters=epochs`?

**Comment by YichengDWu** [comment]
Can you try putting everything on the CPU and see if it works?

**Comment by ba2tripleO** [comment]
I won't be able to run it, the purpose of this pr was to make it use Optimization.jl, if there aren't any issues with that, then merge it. You'll have change the output after running it.

**Comment by YichengDWu** [comment]
I can't change the output since it is not working. I gave a hint to debug it.

**Comment by YichengDWu** [comment]
You will have to change the output since you have it in your hand, if you can make it work on your computer

**Comment by ba2tripleO** [comment]
What is the error you're getting?

**Comment by YichengDWu** [comment]
Look it up in the review comments. I thought you had resolved it, but you just ignored it

**Comment by YichengDWu** [comment]
Also, you don't reassign a value to a global variable in a function without declaring it 
```julia
julia> a = 1
1

julia> function foo()
           @show a
           a = 2
           return nothing
       end
foo (generic function with 1 method)

julia> foo()
ERROR: UndefVarError: a not defined
Stacktrace:
 [1] macro expansion
   @ .\show.jl:1047 [inlined]
 [2] foo()
   @ Main .\REPL[12]:2
 [3] top-level scope
   @ REPL[13]:1
```
If you are not even aware of this, I'm curious how you got it running on your computer

**Comment by ba2tripleO** [comment]
As I said, I proposed the changes needed to make it use Optimization.jl, not to run it, I've got my own projects to work on, it's your project to make this doc example work, so you have to fix any minor issues with it. If you don't want these changes in your pr we can close this.

**Comment by YichengDWu** [comment]
This is not my project. I opened my pr upon Chris's request. It is your organization's responsibility to make it work if you are a member of it. It is Chris who told you to transfer it to Optimization. I assume it is your job. If you are just a random guy I'm really sorry. I don't need to fix nothing since I always make sure my code runs before I submit a PR.

**Pull request closed without merging**